data/src/main/java/com/linkedin/data/codec/entitystream/AbstractDataDecoder.java

```diff
@@ -0,0 +1,361 @@
+      processTokens();
+    }
+    catch (IOException e)
+    {
+      handleException(e);
+    }
+  }
+
+  private void processTokens()
+  {
+    try
+    {
+      DataParser.Token token;
+      while ((token = _parser.nextToken()) != null)
+      {
+        switch (token)
+        {
+          case START_OBJECT:
+            validate(START_OBJECT);
+            // If we are already filling out a DataMap, we cannot reuse _currDataMapBuilder and thus
+            // need to create a new one.
+            if (_currDataMapBuilder.inUse())
+            {
+              _currDataMapBuilder = new DataMapBuilder();
+            }
+            _currDataMapBuilder.setInUse(true);
+            push(_currDataMapBuilder, false);
+            break;
+          case END_OBJECT:
+            validate(END_OBJECT);
+            pop();
+            break;
+          case START_ARRAY:
+            validate(START_ARRAY);
+            push(new DataList(), true);
+            break;
+          case END_ARRAY:
+            validate(END_ARRAY);
+            pop();
+            break;
+          case STRING:
+            validate(STRING);
+            if (!_isCurrList && _currField == null) {
+              _currField = _parser.getString();
+              _expectedTokens = VALUE;
+            } else {
+              addValue(_parser.getString());
+            }
+            break;
+          case INTEGER:
+            addValue(_parser.getIntValue());
+            break;
+          case LONG:
+            addValue(_parser.getLongValue());
+            break;
+          case FLOAT:
+            addValue(_parser.getFloatValue());
+            break;
+          case DOUBLE:
+            addValue(_parser.getDoubleValue());
+            break;
+          case BOOL_TRUE:
+            validate(BOOL_TRUE);
+            addValue(Boolean.TRUE);
+            break;
+          case BOOL_FALSE:
+            validate(BOOL_FALSE);
+            addValue(Boolean.FALSE);
+            break;
+          case NULL:
+            validate(NULL);
+            addValue(Data.NULL);
+            break;
+          case NOT_AVAILABLE:
+            readNextChunk();
+            return;
+          default:
+            handleException(new Exception("Unexpected token " + token + " from data parser"));
+        }
+      }
+    }
+    catch (IOException e)
+    {
+      handleException(e);
+    }
+  }
+
+  protected final void validate(DataParser.Token token)
+  {
+    if (!_expectedTokens.contains(token))
+    {
+      handleException(new Exception("Expecting " + _expectedTokens + " but got " + token));
+    }
+  }
+
+  private void push(DataComplex dataComplex, boolean isList)
+  {
+    if (!(_isCurrList || _stack.isEmpty()))
+    {
+      _currFieldStack.push(_currField);
+      _currField = null;
+    }
+    _stack.push(dataComplex);
+    _isCurrList = isList;
+    updateExpected();
+  }
+
+  @SuppressWarnings("unchecked")
+  private void pop()
```

Comment:
This should be protected. Protobuf for example can do this much more efficiently by not using DataMapBuilder since it is a length prefixed encoding format.

Reply:
I thought about this but protodecoder overriding this would mean private stacks and other variables internal to AbstractDecoder would have to be exposed and introduce the potential risk of error.

Another approach was to maintain size stack in ProtobufParser and use it to return END objects but size would still need to be exposed to Abstract path since it required at initiating complex objects with know size/capacity.

therefore adding size check branch in abstract code path would be better. Abstract methods will call parser getSize if object tokens are returned and if not invalid use it.
Since it adds only a if check on int in main code path for all types of decoder it should be fine as could also be used by potential other future codecs.

Let me know if it looks right, we can discuss further.

Reply:
Add size optimization implementation in this PR only for context